token/services/interop/exchange/wallet.go

```diff
@@ -0,0 +1,142 @@
+			continue
+		}
+		if owner.Type == ScriptTypeExchange {
+			script := &Script{}
+			if err := json.Unmarshal(owner.Identity, script); err != nil {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			if script.Sender.IsNone() {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			logger.Debugf("token [%s,%s,%s,%s] contains a script? Yes", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+
+			// is this expired and I am the sender?
+			now := time.Now()
+			logger.Debugf("[%v]<=[%v] and sender [%s]?", script.Deadline, now, script.Sender.UniqueID())
+			if script.Deadline.Before(now) && w.wallet.Contains(script.Sender) {
+				logger.Debugf("[%v]<=[%v] and sender [%s]? Yes", script.Deadline, now, script.Sender.UniqueID())
+				res = append(res, tok)
+			}
+			logger.Debugf("[%v]<=[%v] and sender [%s]? No", script.Deadline, now, script.Sender.UniqueID())
+		}
+	}
+	return res, nil
+}
+
+// ListByPreImage returns a list of tokens with a matching preimage
+func (w *OwnerWallet) ListByPreImage(preImage []byte) ([]*token2.UnspentToken, error) {
+	unspentTokens, err := w.queryService.ListUnspentTokens()
+	if err != nil {
+		return nil, errors.Wrap(err, "token selection failed")
+	}
+	logger.Debugf("[%d] unspent tokens found", len(unspentTokens.Tokens))
+	var res []*token2.UnspentToken
+	for _, tok := range unspentTokens.Tokens {
+		logger.Debugf("token [%s,%s,%s,%s] contains a script?", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+
+		owner, err := identity.UnmarshallRawOwner(tok.Owner.Raw)
+		if err != nil {
+			logger.Debugf("Is Mine [%s,%s,%s]? No, failed unmarshalling [%s]", view.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, err)
+			continue
+		}
+		if owner.Type == ScriptTypeExchange {
+			script := &Script{}
+			if err := json.Unmarshal(owner.Identity, script); err != nil {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			if script.Sender.IsNone() {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			logger.Debugf("token [%s,%s,%s,%s] contains a script? Yes", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+
+			hash := script.HashInfo.HashFunc.New()
+			if _, err = hash.Write(preImage); err != nil {
+				return nil, err
+			}
+			h := hash.Sum(nil)
+			h = []byte(script.HashInfo.HashEncoding.New().EncodeToString(h))
+
+			logger.Debugf("searching for script matching (pre-image, image) = (%s,%s)",
+				base64.StdEncoding.EncodeToString(preImage),
+				base64.StdEncoding.EncodeToString(h),
+			)
```

Comment:
@adecaro this is a good example where we would like to use `if logger.IsEnabledFor(zapcore.DebugLevel)` to prevent unnecessary base64 encoding.